# Проект на SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

###  Цели исследования  
    Проанализировать базу данных для сформулирования предложений для нового продукта.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

Загрузим требуемые библиотеки для работы и создадим подключение к базе данных

In [1]:
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

### Выведем первые строки каждой таблицы базы данных

Таблица books Содержит данные о книгах:

In [3]:
query = ''' SELECT *
            FROM books
        '''

In [4]:
pd.io.sql.read_sql(query, con = engine).head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
pd.io.sql.read_sql(query, con = engine).info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Таблица authors содержит данные об авторах:

In [6]:
query = ''' SELECT *
            FROM authors
        '''

In [7]:
pd.io.sql.read_sql(query, con = engine).head() 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
pd.io.sql.read_sql(query, con = engine).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Таблица publishers содержит данные об издательствах:

In [9]:
query= ''' SELECT *
            FROM publishers
        '''

In [10]:
pd.io.sql.read_sql(query, con = engine).head() 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
pd.io.sql.read_sql(query, con = engine).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица ratings содержит данные о пользовательских оценках книг:

In [12]:
query = ''' SELECT *
            FROM ratings
        '''

In [13]:
pd.io.sql.read_sql(query, con = engine).head() 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [14]:
pd.io.sql.read_sql(query, con = engine).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица reviews cодержит данные о пользовательских обзорах на книги:

In [15]:
query = ''' SELECT *
            FROM reviews
        '''

In [16]:
pd.io.sql.read_sql(query, con = engine).head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [17]:
pd.io.sql.read_sql(query, con = engine).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Задания

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [18]:
query1 = ''' SELECT COUNT(DISTINCT(book_id))
             FROM books
             WHERE CAST(publication_date AS date) > '2000-01-01'
        '''

In [19]:
count_book = pd.io.sql.read_sql(query1, con = engine)
count_book = count_book.loc[0, 'count']
count_book

819

In [20]:
print(f'После 1 января 2000 года вышло {count_book} книг.')

После 1 января 2000 года вышло 819 книг.


### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [39]:
query2 = ''' SELECT b.title as title,
                    b.book_id,
                    COUNT(DISTINCT(rw.review_id)) AS count_reviews,
                    TRUNC(AVG(rt.rating), 2) AS mean_rating
             FROM books AS b
             LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
             LEFT JOIN reviews AS rw ON b.book_id=rw.book_id
             GROUP BY title, 
                      b.book_id
             ORDER  BY count_reviews DESC
        '''

In [40]:
book = pd.io.sql.read_sql(query2, con = engine)
book

title  book_id  \
0                              Twilight (Twilight  #1)      948   
1                                  Water for Elephants      963   
2                                     The Glass Castle      734   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302   
4    The Curious Incident of the Dog in the Night-Time      695   
..                                                 ...      ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel       83   
996                            The Natural Way to Draw      808   
997   The Cat in the Hat and Other Dr. Seuss Favorites      672   
998                          Essential Tales and Poems      221   
999  Disney's Beauty and the Beast (A Little Golden...      191   

     count_reviews  mean_rating  
0                7         3.66  
1                6         3.97  
2                6         4.20  
3                6         4.41  
4                6         4.08  
..             ...          ...  
995              0         3.66  
996              0         3.00  
997              0         5.00  
998              0         4.00  
999              0         4.00  

[1000 rows x 4 columns]

In [23]:
query_mean_rating = ''' 
SELECT COUNT(f.mean_rating)
FROM
    (SELECT b.title as title,
            COUNT(rw.review_id) AS count_reviews,
            TRUNC(AVG(rt.rating), 2) AS mean_rating
     FROM books AS b
     LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
     LEFT JOIN reviews AS rw ON b.book_id=rw.book_id
     GROUP BY title, 
              b.book_id     
     ORDER  BY mean_rating DESC) as f
GROUP BY f.mean_rating          
HAVING f.mean_rating = 5
        '''

In [24]:
pd.io.sql.read_sql(query_mean_rating, con = engine)

count
0     44

Книга с самым большим количеством обзоров - Twilight (Twilight #1) - 7 обзоров.
44 книги получили наивысшую среднюю оценку - 5 баллов.

In [25]:
query100 = ''' 
SELECT SUM(c.count_reviews)
FROM 
    (SELECT b.title as title,
            COUNT(DISTINCT(rw.review_id)) AS count_reviews,
            TRUNC(AVG(rt.rating), 2) AS mean_rating
            FROM books AS b
            LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
            LEFT JOIN reviews AS rw ON b.book_id=rw.book_id
            GROUP BY title, 
                    b.book_id
            ORDER  BY count_reviews DESC) as c
        '''

In [26]:
pd.io.sql.read_sql(query100, con = engine)

sum
0  2793.0

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [27]:
query3 = ''' SELECT p.publisher AS publisher,
                    COUNT(b.book_id) AS count_book
             FROM publishers AS p    
             INNER JOIN books AS b ON p.publisher_id=b.publisher_id
             WHERE b.num_pages > 50
             GROUP BY publisher
             ORDER BY count_book DESC
             LIMIT 1
        '''

In [28]:
max_publisher = pd.io.sql.read_sql(query3, con = engine)
max_publisher

publisher  count_book
0  Penguin Books          42

In [29]:
max_publisher_name = max_publisher.loc[0,'publisher'] 

In [30]:
print(f'Издательство {max_publisher_name} выпустило наибольшее количество книг толще 50 страниц.')

Издательство Penguin Books выпустило наибольшее количество книг толще 50 страниц.


### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [31]:
query4 = ''' SELECT a.author AS author,
                    AVG(r.rating) AS rating
             FROM authors AS a    
             INNER JOIN books AS b ON a.author_id=b.author_id
             INNER JOIN ratings AS r ON b.book_id=r.book_id
             GROUP BY author
             HAVING COUNT(rating) > 50
             ORDER BY rating DESC
             LIMIT 1          
        '''

In [32]:
top_author = pd.io.sql.read_sql(query4, con = engine)
top_author

author    rating
0  J.K. Rowling/Mary GrandPré  4.288462

In [33]:
name_top_author = top_author.loc[0, 'author']

In [34]:
print(f'{name_top_author} - автор с самой высокой оценкой книг.')

J.K. Rowling/Mary GrandPré - автор с самой высокой оценкой книг.


### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [35]:
query5 = ''' SELECT TRUNC(AVG(list_user.count_reviews), 1) AS mean_count_reviews
             FROM
                (SELECT username,
                        COUNT(review_id) AS count_reviews
                FROM reviews
                WHERE username IN (SELECT username
                                   FROM ratings
                                   GROUP BY username
                                   HAVING COUNT(rating_id) > 50)
                GROUP BY username) AS list_user
          ''' 

In [36]:
mean_count_reviews = pd.io.sql.read_sql(query5, con = engine)
mean_count_reviews

mean_count_reviews
0                24.3

In [37]:
count_reviews = mean_count_reviews.loc[0, 'mean_count_reviews']
count_reviews

24.3

In [38]:
print(f'Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно {count_reviews}.')

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24.3.


## Выводы:  
- После 1 января 2000 года вышло ***819 книг***;
- Для каждой книги было посчитано количество обзоров и средняя оценка;
 - Книга с самым большим количесвтом обзоров - Twilight (Twilight #1) - 7 обзоров.
 - 44 книги получили наивысшую среднюю оценку - 5 баллов.
- Издательство ***Penguin Books*** выпустило наибольшее количество книг толще 50 страниц;
- ***J.K. Rowling/Mary GrandPré*** - автор с самой высокой оценкой книг;
- Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно ***24.3***.